### PR TITLE
Add approved scope to #users

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

**Comment by lowjoel** [comment]
No, I'm against this. It would not be possible to remove the scope from the association. The same reason why I'm trying not to use a default scope anywhere.

**Comment by allenwq** [comment]
So solutions ? It will be very easy for programmers to make mistakes then.

**Comment by lowjoel** [comment]
None. Have to deal with it.

**Comment by allenwq** [comment]
I'm not too agree with you. why we have a default scope on course , other
than manually deal with it ?

**Comment by lowjoel** [comment]
There's no default scope on course.

**Comment by allenwq** [comment]
The courses are scoped to current instance by default.

**Comment by lowjoel** [comment]
That's an abstraction, and code should NOT be dealing with another instance's data.

**Comment by allenwq** [comment]
@lowjoel Updated

File: app/models/concerns/course_user_concern.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module CourseUserConcern
```
Comment (lowjoel) [inline app/models/concerns/course_user_concern.rb]: CourseUser::UsersConcern

File: spec/models/course_spec.rb [hunk 2]
```diff
@@ -167,5 +167,15 @@
         expect(subject.registration_key).to start_with('C')
       end
     end
+
+    describe '#users' do
+      let(:course) { create(:course, creator: owner, updater: owner) }
+      let(:un_approved_user) { create(:course_user, course: course).user }
+      let(:approved_user) { create(:course_user, :approved, course: course).user }
+      subject { course.users.with_approved_state }
```
Comment (lowjoel) [inline spec/models/course_spec.rb]: subject mismatch.

**Comment by lowjoel** [comment]
You probably should change the title, and include the points that we discussed this morning (while we don't directly change the association)

File: spec/models/course_spec.rb (hunk 2 above)
Comment (allenwq) [inline spec/models/course_spec.rb]: why mismatch ?

Comment (lowjoel) [inline spec/models/course_spec.rb]: you are describing `#with_approved_state`, not #users.

Comment (allenwq) [inline spec/models/course_spec.rb]: Shall I nest `#with_approved_state` under `#users` ?  Because these two functions are supposed to work together.

Comment (lowjoel) [inline spec/models/course_spec.rb]: yes, you should.

**Comment by allenwq** [comment]
Things we discussed this morning:

1. Don't add a default scope directly, it will not be possible to remove and easily break other code like abilities

2. Deal with this by adding new non-default scopes in concerns , call `course.users.with_approved_state` to get approved users instead.

3. The user registrations are complicated, programmers need to understand and deal with it.

File: spec/models/course_spec.rb [hunk 3]
```diff
@@ -167,5 +167,17 @@
         expect(subject.registration_key).to start_with('C')
       end
     end
+
+    describe '#users' do
+      let(:course) { create(:course, creator: owner, updater: owner) }
+      describe '#with_approved_state' do
+        let(:un_approved_user) { create(:course_user, course: course).user }
```
Comment (lowjoel) [inline spec/models/course_spec.rb]: oops, should be `unapproved_user`

Comment (allenwq) [inline spec/models/course_spec.rb]: updated, thanks !

File: spec/models/course_spec.rb [hunk 4]
```diff
@@ -167,5 +167,17 @@
...
+
+    describe '#users' do
+      let(:course) { create(:course, creator: owner, updater: owner) }
+      describe '#with_approved_state' do
+        let(:unapproved_user) { create(:course_user, course: course).user }
+        let(:approved_user) { create(:course_user, :approved, course: course).user }
+        subject { course.users.with_approved_state }
+
+        it { is_expected.to include(approved_user) }
+        it { is_expected.not_to include(un_approved_user) }
```
Comment (lowjoel) [inline spec/models/course_spec.rb]: ...here also lol

Comment (allenwq) [inline spec/models/course_spec.rb]: oops..

**Comment by lowjoel** [comment]
Should we put a comment in the association declaration?

**Comment by allenwq** [comment]
Comments added, any comments about it ?

File: app/models/course.rb [hunk 5]
```diff
@@ -13,7 +13,11 @@ class Course < ActiveRecord::Base
 
   belongs_to :creator, class_name: User.name
   has_many :course_users, inverse_of: :course, dependent: :destroy
-  has_many :users, through: :course_users
+  # Returns all the users related to the course regardless of course_user state.
```
Comment (lowjoel) [inline app/models/course.rb]: Use `@!attribute [r] users` to describe the `has_many` association.

**Comment by lowjoel** [comment]
Nice!

**Pull request merged**